### ISSUE TITLE
DNS resolution not working when IPv6 is enabled only

### ISSUE THREAD

**Issue by quater** [lxc/lxd#4088]
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.3 LTS"
``` 
```
lxc info
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFXDCCA0SgAwIBAgIRANMRqupUh1oSkO5hiMAdY9QwDQYJKoZIhvcNAQELBQAw
    OzEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEbMBkGA1UEAwwScm9vdEB1
    YnVudHUteGVuaWFsMB4XDTE3MTIwODE5NDk1NloXDTI3MTIwNjE5NDk1NlowOzEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEbMBkGA1UEAwwScm9vdEB1YnVu
    dHUteGVuaWFsMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAv398y/h9
    lM9pkHnTDqiYg7qxqyIC0zDuAtETp1hndFzh9jYR02MrW7YvThNX2SYDrzKBDAyL
    TLtvfYJhnDT/o/8gTbel4CUC0aBV7gSf94cQu3IbTTZwhRZJvVS5Ck2ObFX5gwaG
    UUsxlHFy8TV2maZDcI2s8D78Wt4LjO43Xf+vzU0s0iYoIRbW4/uAXixNRuqHtfAp
    RN66Z9LmPIlb8jn2hIXGICvt0ILxsuYlC+Vcrf4hPqIkNZCnMbD1wBKnZeB8tyen
    zRGYqgD4qjvzM9ksNEJ1RFkhla7YhrzcOlztf5QYmgWF37o6/QkDU0FLdcJm7MLt
    6D6zBu9kptzcG78PVvAkWDVZLX9I4oMIfTpSjIm/um4ejpnqkasKeUnLA/geSIog
    jRq3q5ACekrQKtFh+/MoDg9iOlmh7cfIQ642tWaidsz7VNsoCrsKbGXfomI+746k
    glAiQ26PT8pDy3pj7CwlYmSgwv40QC4jcVW5Jjv4pA4gIOXXjS+TekXPfWX1Xjx2
    wnR5JWN4nnRI8bDOY4EtQLr5m1PebBpTM8y/MYJ8EAKk7ceKw0puNtdDQlam1oqb
    yI68dDP08BTOD3UdULjhpgkiSdn1T9eZ4K6/8wrIM45Oj4BEBZI5/7lqflZtqQZk
    xxyCNDRRfgb3/ghF5syiA7cPJ1abGdB1+MUCAwEAAaNbMFkwDgYDVR0PAQH/BAQD
    AgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwJAYDVR0RBB0w
    G4INdWJ1bnR1LXhlbmlhbIcECgACD4cErByABjANBgkqhkiG9w0BAQsFAAOCAgEA
    T0w9cH1VAVazxaVvUHTAeG7AREKWI3czIPnoemyfSHBk9B8wzflxtqtM9LaSDcsL
    PveAHheyxknaW1fM9dPd163rOv8G3Q+wFSMFdXzfC1+gWXBH23gVekd5xcxGrkhR
    erao74H+rdOx5xO9WL49b3fFRVA7NUNmpGXAXBd4kd61bF1CB9mywH4tIs5fprjz
    dvcIXNVabepqT2TtGln66/S/81qzxfk77Uy6HBx048MRiKfa0s9Zqav58NJ4kqfS
    ExqfN1dPgLAKORtNyUGn10+Hqo0EdU6/j+JGPIsK72MqzLBla6PDEcBY2WFijrqG
    AEqUr7hB94tuDuyDAMnAFN2LPxZ+Dg3N5R1WTP1eIfOHomowKLc+FXHYIdkZtux3
    w/7DuY6BAHabSz/gBK9uTDvuKaanOFOEzhQL5HpRZYS4PqWLEHQZDRR+Foh6u2BA
    qX3b7ZEVo/nFm7IOYMDqPClxnh58sVN+zInJNSHtcdaceowJ/bS2BrMRT995ID+9
    5YcjbBl/DjbDdVDhJnxcRYyowh61A090B34npGSn4u2WZmmQkTQQROwhAjuU5oLE
    hT+P8UUwrhIuOyLWL1Rf6qcmCdO4NZuzO+vVIzlo93NnTbFC0CTZtDdnxzOt7skz
    vfVgbYBss2NJIiKyu/KVyhgJACW6r4AZcefcqVD6Itw=
    -----END CERTIFICATE-----
  certificate_fingerprint: c702735c4def66070c69707d74d0bdf432198ffa6a7eae7c52bf1332da72d2a0
  driver: lxc
  driver_version: 2.1.1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-98-generic
  server: lxd
  server_pid: 2747
  server_version: "2.20"
  storage: btrfs
  storage_version: "4.4"
```
# Issue description

DNS resolution not working when IPv6 is enabled only.

# Steps to reproduce
```
## Not working when only IPv6 is enabled
lxc network create br-b ipv4.address=none ipv6.address=auto ipv6.dhcp=true
lxc init ubuntu:16.04 b1
lxc config device add b1 eth0 nic nictype=bridged parent=br-b name=eth0
lxc start b1
lxc init ubuntu:16.04 b2
lxc config device add b2 eth0 nic nictype=bridged parent=br-b name=eth0
lxc start b2
lxc exec b1 -- /bin/bash
==>>
root@b1:~# ping6 b2
unknown host
```

```
## Working when both, IPv4 and IPv6 are enabled
lxc network create br-a ipv6.dhcp=true
lxc init ubuntu:16.04 a1
lxc config device add a1 eth0 nic nictype=bridged parent=br-a name=eth0
lxc start a1
lxc init ubuntu:16.04 a2
lxc config device add a2 eth0 nic nictype=bridged parent=br-a name=eth0
lxc start a2
lxc exec a1 -- /bin/bash
==>>
root@a1:~# ping6 a2
PING a2(a2.lxd) 56 data bytes
64 bytes from a2.lxd: icmp_seq=1 ttl=64 time=0.240 ms
64 bytes from a2.lxd: icmp_seq=2 ttl=64 time=0.120 ms
64 bytes from a2.lxd: icmp_seq=3 ttl=64 time=0.127 ms
```

**Comment by stgraber** [comment]
That's expected and not a LXD bug.

By default LXD will provide IPv6 router advertisements including RDNSS and DNSSL which are sufficient to configure IPv6 without any DHCP if the system knows how to parse those router advertisements. Additionally, LXD also provides stateless DHCPv6 which also provides the DNS server and search domains (similar to RDNSS and DNSSL).

The catch is that most Linux distributions do neither of those things by default. They just get the auto-configured IPv6 address from the kernel and that's it. When you have IPv4 enabled, they'll do DHCP on IPv4 and get a working DNS server through that.

So to get your container working with IPv6 only using the stock LXD config, you either need:
 - Something like NetworkManager which will extra the RDNSS and DNSSL data from the RA
 - Run dhclient in stateless mode to get the DNS and search domain information over stateless DHCPv6

The latter tends to be easier and can be done by modifying your network configuration inside your Ubuntu container to look like:
````
auto eth0
iface eth0 inet6 auto
    dhcp 1
````

This particular part will vary from distro to distro as they use different network management tools.

**Comment by stgraber** [comment]
Note that you can also enable stateful DHCPv6 in LXD by setting ipv6.dhcp.stateful=true on the bridge, at which point you'd need a stateful DHCPv6 client in the container. For Ubuntu that's done with:

```
auto eth0
iface eth0 inet6 dhcp
```

**Comment by quater** [comment]
@stgraber Thank you for the clarification and the useful information. Just adding this comment for completeness.

Based on your response I ended up using the `ipv4.address=none` and `ipv6.dhcp.stateful=true` options.

However, the only way I got it to actually resolve DNS queries with Ubuntu 16.04, was by uninstalling the `isc-dhcp-client` and `isc-dhcp-common` packages and then installing `dhcpcd5`. With that the `/etc/network/interface` configuration files became obsolete and things had to be configured in the `/etc/dhcpcd5.conf` file.

To ensure that `dhcpcd5` only handles IPv6 for the eth0 interface, I had to add
```
interface eth0
  ipv6only
```
to the `/etc/dhcpcd5.conf` file.
https://manpages.debian.org/jessie/dhcpcd5/dhcpcd.conf.5.en.html

With that in place, IPv6 addresses where assigned through DHCPv6 and also where discoverable via DNS.
